Run Collect-info through Edgeview

naiming-zededa opened the pull request:
there are a number of request to have edgeview support for collectinfo,
this patch has the list of implementation:

- Add the implementation on running collect-info through Edgeview
- Egeview is mostly only with read permission access to device filesystems
   and it does not want to import all the pkgs and replicate the collect-info.sh
   operations and maintain compatibility of two sets of the collect-info
- this implementation has a protocol to setup a request for debug container
  to start run 'collect-info.sh', and remove the request file after it is done;
  Edgeview then uses existing copy file protocol to transfer the collect-info
  tarball file back to user's laptop; then Edgeview put another request for
  debug container to remove the created tarball file in /persist
- debug container has added a background task to monitor the requests
- Since collect-info is mostly live info, there is no need to get all the newlogs,
  only fetch the newlogs of the past 10 days.
- Introduced debug/debug-services.sh around the exiting sshd and the new edgeview-collectinfo.sh

milan-zededa commented on pkg/debug/scripts/collect-info.sh:
```diff
@@ -20,6 +20,7 @@ INFO_DIR="eve-info-v$VERSION-$DATE"
 TARBALL_FILE="/persist/$INFO_DIR.tar.gz"
 SCRIPT_DIR=$(dirname "$(readlink -f "$0")")
 
+READ_LOGS_DAYS=
```
Please increase the VERSION variable after making changes in this script.
(probably to 26 if we merge this sooner: https://github.com/lf-edge/eve/pull/4095/files)

naiming-zededa replied:
I thought i did, but i guess someone upgraded before this pr, and i missed that. will do.

eriknordmark replied:
You'll need to increase it again since I merged a PR to collect-info.sh today

naiming-zededa replied:
ok. Changed.

milan-zededa commented on pkg/debug/ssh.sh:
```diff
@@ -7,6 +7,43 @@ echo -1 > /proc/sys/kernel/perf_event_paranoid
+            echo "edgeview request to run collect-info..." > /dev/kmsg
+            # the newlog part of the collection, only collect last 10 days
+            /usr/bin/collect-info.sh -t 10
+            # romove the request file
```
romove -> remove

naiming-zededa replied:
ok.

milan-zededa commented on pkg/edgeview/src/system.go:
```diff
@@ -1306,6 +1315,140 @@ func runTechSupport(cmds cmdOpt, isLocal bool) {
+	}
+	defer watcher.Close()
+
+	done := make(chan bool)
```
How about changing channel value to error and then returning the error from this function?
This would preserve errors like "Timeout exceeded", which could be printed to the console instead of potentially confusing message "can not find the eve info file".

naiming-zededa replied:
sure.

christoph-zededa commented on pkg/debug/Dockerfile:
```diff
@@ -9,14 +9,14 @@
 # into the package: see abuild/etc/abuild.conf.
 FROM lfedge/eve-recovertpm:0da17f65aba4fb09c83944cf5847dd7b523118b4 as recovertpm
-FROM lfedge/eve-alpine:1f7685f95a475c6bbe682f0b976f12180b6c8726 as build
+FROM lfedge/eve-alpine:43ba9262f1b3c71b7a501b549d0cfb811a2d81ee as build
```
Please keep it consistent and also bump it everywhere else.

christoph-zededa replied:
You can use something like
```
find ./ -name Dockerfile -exec perl -pni.bak -e 's/lfedge\/eve-alpine(\S+)/lfedge\/eve-alpine:43ba9262f1b3c71b7a501b549d0cfb811a2d81ee/'
 "{}" \;
```
to make your life a bit easier

naiming-zededa replied:
do we do this across the EVE now, i thought we still selectively bump up? will search for those. thanks.

christoph-zededa commented on pkg/debug/scripts/collect-info.sh:
```diff
@@ -47,11 +48,12 @@ usage()
     echo "Read-logs mode:"
     echo "       -d                   - read device logs only"
     echo "       -a APPLICATION-UUID  - read specified application logs only"
+    echo "       -t NUMBER-OF-DAYS    - read logs from the last NUMBER-OF-DAYS [1-30]"
```
Why is it capped at 30?

naiming-zededa replied:
just a number, i thought if we are that far way, why not just get all the logs

christoph-zededa commented on pkg/debug/ssh.sh:
```diff
@@ -7,6 +7,43 @@ echo -1 > /proc/sys/kernel/perf_event_paranoid
 KEYS=$(find /etc/ssh -name 'ssh_host_*_key')
 [ -z "$KEYS" ] && ssh-keygen -A >/dev/null 2>/dev/null
 
+# Background monitoring non-ssh related tasks
```
if it is not ssh related, why is it then in the `ssh.sh` script? Perhaps this script should then be renamed or another one should be created?

rene replied:
I agree with @christoph-zededa , maybe this could be part of edgeview container, you can use something like `eve exec debug ....` to run collect-info related commands within debug container...

christoph-zededa replied:
Yes (https://github.com/lf-edge/eve/pull/4106#discussion_r1691321362 )

but better not to use `eve exec debug ...` and rather use the containerd interface or `setns` to have better error handling

naiming-zededa replied:
maybe i can change this into 'debug_operations.sh' ?
1) for edgeview code, I had tried to remove all the shell commands due to previous review comments
2) try not do have the edgeview to gain all the capabilities of the debug, such as all the container related operations, 'ctr', etc.
3) the protocol here is quite simple operation actually, deposit a file, monitor the file for a specific task job

christoph-zededa replied:
> try not do have the edgeview to gain all the capabilities of the debug

Can you explain more?

F.e. if you use `setns/fork` then it would not be edgeview that gains all the capabilities of the debug container.

> the protocol here is quite simple

Well, it is simple as it lacks:
* error propagation
* properly handling of a second request that comes in parallel

naiming-zededa replied:
> > try not do have the edgeview to gain all the capabilities of the debug
> 
> Can you explain more?
> 
> F.e. if you use `setns/fork` then it would not be edgeview that gains all the capabilities of the debug container.
> 
> > the protocol here is quite simple
> 
> Well, it is simple as it lacks:
> 
>     * error propagation
> 
>     * properly handling of a second request that comes in parallel

@christoph-zededa  edgeview is for remote access, so it is more security sensitive, currently only allow to write into /run/edgeview, if we enable the setns, then we can ask debug or other containers to remove or modify other files, and the it also needs to import all the container related packages into the edgeview container.

to implement container task management into edgeview in golang is another way, but that is too much
work for the purpose of just wanting to run 'collect-info'.

One thing I probably would do is the remove this inotifywait function from the task loop, a simple
sleep of 5 seconds then check file existance seems reasonable.

christoph-zededa replied:
`setns` is disabled? Where can I see this?
Perhaps you want to add an apparmor profile then, too :-)

Anyways renaming it to `debug_operations.sh` or perhaps `init.sh` is fine for me.

I didn't like that you invent a new IPC protocol here, but as others are doing the same, I guess it is ok.

naiming-zededa replied:
I have renamed this to 'debug_tasks.sh' already in the latest version of the commit.
This is only a one-off operation for edgeview (need the execution from another container). If later we found more cases, we can revisit to 'apparmor' :-) or more general way.

christoph-zededa commented on pkg/debug/ssh.sh:
```diff
@@ -7,6 +7,43 @@ echo -1 > /proc/sys/kernel/perf_event_paranoid
+        if [ -f "/run/edgeview/edgeview-request-collect-info" ]; then
+            echo "edgeview request to run collect-info..." > /dev/kmsg
+            # the newlog part of the collection, only collect last 10 days
+            /usr/bin/collect-info.sh -t 10
```
Where will the output (stdout, stderr) of `collect-info.sh` be written to?

naiming-zededa replied:
will get collected through device newlogs

christoph-zededa commented on pkg/edgeview/src/system.go:
```diff
@@ -1306,6 +1315,140 @@ func runTechSupport(cmds cmdOpt, isLocal bool) {
+				if !ok {
+					return
+				}
+				if event.Op&fsnotify.Remove == fsnotify.Remove {
+					if event.Name == filePath {
```
```suggestion
				if event.Op&fsnotify.Remove == fsnotify.Remove && event.Name == filePath {
```

christoph-zededa commented on pkg/edgeview/src/system.go:
```diff
@@ -1306,6 +1315,140 @@ func runTechSupport(cmds cmdOpt, isLocal bool) {
+				if !ok {
+					return
+				}
+				log.Noticef("monitorFileDeletion:: error: %v", err)
```
```suggestion
				log.Noticef("monitorFileDeletion: error: %v", err)
```

christoph-zededa commented on pkg/edgeview/src/system.go:
```diff
@@ -1306,6 +1315,140 @@ func runTechSupport(cmds cmdOpt, isLocal bool) {
+				log.Noticef("monitorFileDeletion:: error: %v", err)
+			case <-time.After(timeout - time.Since(startTime)):
+				log.Noticef("monitorFileDeletion: Timeout exceeded")
+				done <- true
```
```suggestion
				done <- false
```
Shouldn't it be `false` because it is not done?

naiming-zededa replied:
this is timeout, so it's done.

christoph-zededa commented on pkg/edgeview/src/system.go:
```diff
@@ -1306,6 +1315,140 @@ func runTechSupport(cmds cmdOpt, isLocal bool) {
+
+		// Check if the file was modified after the create time
+		if fileInfo.ModTime().After(createTime) {
+			return match, nil // Return the first recent file found
```
What happens if there is more than one match?

naiming-zededa replied:
there should not be any file matches the pattern of collect-info tarball, and the timestamp is after we started collection.

christoph-zededa replied:
What if the node crashes and reboot during creation, then there can be two, can't there?

milan-zededa replied:
The file is in /run, i.e. not persisted, so this should not happen. Also the filename is constant.
@naiming-zededa Correct?

christoph-zededa replied:
> The one before the reboot is older than this time.

How will it be cleaned up?

naiming-zededa replied:
no plan to clean is up, like if users manually run collect-info.sh and they may not delete this.

naiming-zededa replied:
one thing we can do is let the debug container to check, if a collect-info file in /persist has been around more than 2 weeks, it can go ahead to delete this file.

naiming-zededa replied:
implemented the task in debug-tasks.sh for removing this type of tarball files if older than 10 days

christoph-zededa replied:
Hmm, I am not sure if that is a good idea, because it also cleans up the collect-info output manually triggered, doesn't it?
Then:
1. 10 days are too short, perhaps a user created a collect-info and wants to copy it later to a usb stick
2. 10 days are too long, perhaps a user expects after 2 hours that the space has been freed again and they can use it for another vm

naiming-zededa replied:
for 1) we'll ask them edgeview can do for them now, no need to log in and trigger then wait for 20 days to port to usb anymore
for 2) it's better than we leave them now unnoticed. since using edgeview, this will not likely to happen as before, mainly is to clean up unexpected cases.

christoph-zededa replied:
But the user might have created the collect-info output not with edgeview or might have some created with edgeview and some on the console.

naiming-zededa replied:
yes, but we either decide to clean them up, if they forget about them, or we don't since we have no idea how long they want to leave them and come back to fetch them later. So, i'm taking a middle point, and think after 10 days, they probably forget about them. Most people would immediately download them but they may forget to delete the one on the device.

christoph-zededa replied:
I fear the following could happen:

User is somewhere out in the field without fast internet connection.
A problem happens.
They call their IT support staff and they tell them (over the phone) to run `collect-info.sh` and reboot.
After three weeks they come home and the IT support wants to extract the collect info output and it is gone.

(The IT support is too lazy to also explain how to copy that file to a usb stick)

naiming-zededa replied:
Ok, I'll remove this task then.

naiming-zededa replied:
@christoph-zededa I changed the task actually, so now it only deletes edgeview generated old tarball files, your IT support guy should be fine now :-)

christoph-zededa replied:
:-)

christoph-zededa commented on pkg/edgeview/src/system.go:
```diff
@@ -1306,6 +1315,140 @@ func runTechSupport(cmds cmdOpt, isLocal bool) {
+	}
+	defer file.Close() // Ensure the file is closed after it's opened/created.
+
+	if content != "" {
+		if _, err := file.WriteString(content); err != nil {
+			return err // Return the error if writing the content fails
+		}
+	}
+
+	return nil
```
```suggestion
	if content == "" {
	    return nil
	 }
	_, err := file.WriteString(content);
	return err // Return the error if writing the content fails

```

christoph-zededa commented on pkg/edgeview/src/system.go:
```diff
@@ -73,6 +80,8 @@ func runSystem(cmds cmdOpt, sysOpt string) {
 			getDownload()
 		} else if strings.HasPrefix(opt, "ps/") || opt == "ps" {
 			runPS(opt)
+		} else if strings.HasPrefix(opt, "collectinfo") {
```
```suggestion
		} else if opt == "collectinfo {
```

christoph-zededa commented on pkg/edgeview/src/system.go:
```diff
@@ -1306,6 +1315,140 @@ func runTechSupport(cmds cmdOpt, isLocal bool) {
+func touchRequestFile(filename, content string) error {
+	file, err := os.OpenFile(filename, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0644)
+	if err != nil {
+		return err
```
```suggestion
		return fmt.Errorf("error opening file %s: %w", filename, err)
```

rene commented on pkg/debug/scripts/collect-info.sh:
```diff
@@ -433,9 +443,17 @@ find /sys/kernel/security -name "tpm*" | while read -r TPM; do
+if [ -n "$READ_LOGS_DAYS" ]; then
+    mkdir -p "$LOG_TMP_DIR"
+    # Find and copy log files from /persist/newlog to $LOG_TMP_DIR in previous days
+    find /persist/newlog -type f -mtime -"$READ_LOGS_DAYS" -exec cp -p {} "$LOG_TMP_DIR" \;
```
The reason to create a link of /persist/newlog it's to avoid coping log files that can be huge, but here you are copying anyways, which can lead the device to run out of disk space. This can be mitigated by creating symbolic links (inside $LOG_TMP_DIR) instead of copy all the files...

naiming-zededa replied:
good point Rene, will do. Although i'm only getting 10 days of logs and it's just few mega bytes here. but yes let me try to use 'ln'.

christoph-zededa commented on pkg/edgeview/src/system.go:
```diff
@@ -1306,6 +1315,140 @@ func runTechSupport(cmds cmdOpt, isLocal bool) {
+// edgeview can also request to remove the tar.gz file by creating
+// /run/edgeview/edgeview-request-remove-tar.gz file, and debug container
+// will remove the generated tar.gz file.
+func runCollectInfo() {
```
Isn't this over-complicated?
Instead of just calling `collect-info.sh` directly, you let a shell script in the debug container check if some files exist and then let it run. What are the advantages of doing it like this? I see a lot of code just for handling this re-invention of an RPC.

naiming-zededa replied:
i think the main thing is the avoid to expend the edgeview into similar capabilities of 'debug' container, has the 'eve enter' capability, etc. i can not do simply by using socket those things also; thus a simple file creation protocol for this specific task for the job.

naiming-zededa commented on pkg/edgeview/src/system.go:
```diff
@@ -1306,6 +1315,140 @@ func runTechSupport(cmds cmdOpt, isLocal bool) {
+// /run/edgeview/edgeview-request-remove-tar.gz file, and debug container
+// will remove the generated tar.gz file.
+func runCollectInfo() {
+	if _, err := os.Stat(requestCollectInfoFile); err == nil {
+		fmt.Printf("runCollectInfo, collect info request already in progress\n")
+		return
+	}
+
+	startTime := time.Now() // Record the start time
+
+	err := touchRequestFile(requestCollectInfoFile, "")
+	if err != nil {
+		fmt.Printf("runCollectInfo, can not create collect info request file\n")
+		return
+	}
```
can i do a rename() operation in this.

christoph-zededa replied:
You're doing now the rename, but still it is not atomic, because first you do `os.Stat` and then you call `touchRequestFile`.

naiming-zededa replied:
Hmm, the os.Stat() is only error condition check, it jumps out in case there is such a file exist, it has nothing to do with the newly touched file being generated. Do i miss something?

christoph-zededa replied:
Isn't the `os.Stat` there to check that no other collect-info is currently running?

What happens if the user starts two at the same time? Then for both the error return value of `os.Stat` will be not nil and they both continue.

naiming-zededa replied:
this will be extremely unlikely event. even it happens, not much problem results: yes, they both continue, which means the same request file is generated and regenerated. the debug task is working on this, and announce the tarfile is done. say both edgeview sessions will download the tarfile. The fast one will finish first and announce we can remove the tarfile, the second one will get an error during still downloading, saying copy file failed. The 2nd guy just have to retry one more time.

christoph-zededa replied:
If you don't want to fix it, I think this text should be added to the code as a comment.

christoph-zededa replied:
Thx. Resolved.

OhmSpectator commented:
Commit messages, pleeeese 😭😭😭

naiming-zededa commented:
> Commit messages, pleeeease 😭😭😭

hmm, i thought the one before were commit messages, but I added something before the list, is that what you meant?

OhmSpectator commented:
> hmm, i thought the one before were commit messages, but I added something before the list, is that what you meant?

No, I mean the messages that you add to any commit when creating it...
Like here: https://github.com/lf-edge/eve/pull/4097/commits/e74fae7063666bc90dab700fe1e40de0a56ab773
Or here: https://github.com/lf-edge/eve/pull/4093/commits/caad6b591cff5eed5be3a3230d0b0ae60e6a1b99

Without proper commit messages, it's hard to understand the changes made when looking through the history or git blame. It's nice you've added a PR description here on GitHub, but it will be lost after the PR is merged: it will be impossible to see it from an IDE or the command line...

naiming-zededa commented:
@OhmSpectator added in the commit. to see if this is good. thanks.

christoph-zededa commented on pkg/debug/debug-tasks.sh:
```diff
@@ -0,0 +1,62 @@
+            echo "edgeview request to run collect-info..." > /dev/kmsg
+            # the newlog part of the collection, only collect last 10 days
+            /usr/bin/collect-info.sh -t 10
+            # romove the request file
```
```suggestion
            # remove the request file
```

naiming-zededa replied:
ok, missed that.

naiming-zededa commented on pkg/debug/debug-tasks.sh:
```diff
@@ -0,0 +1,78 @@
+#!/bin/sh
+
```
ok, added.

naiming-zededa commented on pkg/debug/debug-tasks.sh:
```diff
@@ -0,0 +1,78 @@
+            fileToRemove=$(cat /run/edgeview/edgeview-request-remove-tar-gz)
+            echo "edgeview request to remove $fileToRemove file" > /dev/kmsg
+            if [ -f "$fileToRemove" ]; then
+                rm "$fileToRemove"
```
that is a good point. Changed.

eriknordmark reviewed (commented):
Note that the pathname for the eve-info files changed in https://github.com/lf-edge/eve/pull/4125 so they are in the /persist/eve-info/ directory instead of just in /persist/

So presumably you need to add that eve-info component to the pathnames you use here.

naiming-zededa commented:
> Note that the pathname for the eve-info files changed in #4125 so they are in the /persist/eve-info/ directory instead of just in /persist/
> 
> So presumably you need to add that eve-info component to the pathnames you use here.

missed that. will change.

naiming-zededa commented:
Changed to reflect the new /persist/eve-info/ as the base. I removed the 'mkdir -p "/persist/$INFO_DIR_SUFFIX"' in the collect-info.sh, since this created directory is not used, and never removed after.
I have tested this new change for both manual run and through the edgeview.

eriknordmark commented:
> Changed to reflect the new /persist/eve-info/ as the base. I removed the 'mkdir -p "/persist/$INFO_DIR_SUFFIX"' in the collect-info.sh, since this created directory is not used, and never removed after. I have tested this new change for both manual run and through the edgeview.

Odd, because I had to add that mkdir to make the tar command not fail when I just did a 'eve enter debug' followed by collect-info.sh.
But if you tested that, then there must have been some fluke when I ran it.

naiming-zededa commented:
fixed a yetus complaining

eriknordmark commented on pkg/debug/scripts/collect-info.sh:
```diff
@@ -64,9 +66,20 @@ while getopts "vhsa:dj" o; do
             ;;
+        e)
+            INFO_DIR_SUFFIX="eve-info-edgeview-v$VERSION-$DATE"
+            TARBALL_FILE="/persist/eve-info/$INFO_DIR_SUFFIX.tar.gz"
```
How about moving this duplicated line after the getopts loop.

And it is a bit more readable if the default INFO_DIR_SUFFIX is set just above the getopts loop.

eriknordmark reviewed (commented):
Do we have any automatic tests of edgeview? And how hard would it be to create some basic one to just check that it is functional?

naiming-zededa commented:
> Do we have any automatic tests of edgeview? And how hard would it be to create some basic one to just check that it is functional?

Currently I don't have this infra. We need to create a task/jira item o this.

eriknordmark reviewed (approved):
LGTM

eriknordmark commented:
@naiming-zededa the regression/eden tests passed, but there might be some open comments/issues on the PR. Can you check if those are taken care of to satisfaction?

naiming-zededa commented:
> @naiming-zededa the regression/eden tests passed, but there might be some open comments/issues on the PR. Can you check if those are taken care of to satisfaction?

@eriknordmark I went through them, they should all be resolved.
